### PR TITLE
FPS mouse input, UI fixes, and Movement Netcode updates

### PR DESCRIPTION
**Pull request by Meisaka**
- Looking around with the mouse was made into a toggle (RMB) instead of momentary, this makes it more "playable"
- Click dragging on UI elements is fixed
- A lot of the entity position prediction code has been reviewed and updated, this should make player movement smoother
- Also fixes some nasty bugs
- Clicking on virtual computer consoles to type into them now works as expected

## Description

- The GameStateQueue has been split to separate files between the client and server, most of the previous code ended up in the client. The server doesn't use it as a queue, but a stub version was added to the server anyways to be a bit consistent, it might also be useful for server rollback netcode in the future.
- Header files were added to project generation, mostly for MSVC to list them, hopefully this doesn't cause problems.
- Some debugging info related to movement and netsync has been added to the UI
- Rigid body objects are no longer removed/added every simulation frame
- Physics system uses step-time transforms to save state instead of interpolated ones, needed because we reset them every tick
- Server was changed to use fixed physics time steps directly, this just makes the number of steps more predictable (1 to 1)
- The game state updates now include time stamps, this changes the protocol slightly
- EventQueue got an update and now uses a mutex per queue to stop it from breaking horribly
- The clicking in the console no longer focuses the text box unless you click it directly, this was quite annoying during testing.

## Motivation and Context

- Most FPS style games always have the mouse captured in-play this provides something familiar
- UI elements couldn't be dragged, including selecting text
- Movement was really jumpy and unpleasant to see

## How Has This Been Tested?

Full builds run with MSVC 2017 on Win2008R2 and Clang on MacOS 10.12.6
Server and client were run on both platforms, connected to server both local and remote over a LAN (10 ms ping reported by TEC)
Movement tested by strafing and forward/back and observing subjective "smoothness" (it's okay...)
UI drag fixes tested by typing in the console text bar, then click dragging the text to select some
EventQueue issue found by providing lots of movement inputs over a few minutes, after fix, this doesn't happen.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: client/game.cpp [hunk 1]
```diff
@@ -19,7 +19,11 @@ namespace tec {
 	using networking::ServerMessage;
 	using networking::MessageType;
 
-	Game::Game(std::string config_file_name) :
+	Game::Game(OS &_os, std::string config_file_name) :
+		os(_os),
+		stats(),
+		game_state_queue(this->stats),
+		server_connection(this->stats),
```
Comment (Meisaka) [inline client/game.cpp:26]: These cause warnings and probably have be reordered in another commit

File: client/gui/server-connect.cpp [hunk 2]
```diff
@@ -69,7 +69,27 @@ namespace tec {
...
+		ImGui::Text("States:% 2" PRIu64 "  Cmd In-flight:% 2" PRIu64,
+			connection_stats.server_state_count,
+			connection_stats.current_command_id - connection_stats.current_acked_id);
+		ImGui::Text("   Cmd: % 9" PRIu64 " Ack: % 9" PRIu64,
+			connection_stats.current_command_id,
+			connection_stats.current_acked_id);
+		glm::vec3 pos = connection_stats.client_position;
+		glm::vec3 vel = connection_stats.client_velocity;
+		ImGui::Text("X: %1.3f, %1.3f, %1.3f", pos.x, pos.y, pos.z);
+		ImGui::Text("V: %1.3f, %1.3f, %1.3f", vel.x, vel.y, vel.z);
```
Comment (Meisaka) [inline client/gui/server-connect.cpp:92]: This debugging information could possibly be moved elsewhere, but as a WIP I put it here.

File: client/server-connection.hpp [hunk 3]
```diff
@@ -61,6 +62,11 @@ namespace tec {
 				return this->average_ping;
 			}
 
+			ping_time_t GetEstimatedDelay() {
+				ping_time_t since_recv = std::chrono::duration_cast<std::chrono::milliseconds>(
+					std::chrono::high_resolution_clock::now() - recv_time).count();
+				return (this->stats.estimated_server_time - this->stats.last_state_time) + since_recv;
```
Comment (Meisaka) [inline client/server-connection.hpp:68]: This method is used for displaying the state delay, it's not reliable and currently only for stats display

File: client/voxel-volume.cpp [hunk 4]
```diff
@@ -236,7 +236,7 @@ namespace tec {
 				const Position* pos = Entity(entity_id).Get<Position>();
 				const Orientation* orientation = Entity(entity_id).Get<Orientation>();
 				glm::mat4 model_view = glm::inverse(glm::translate(glm::mat4(1.0), pos->value) * glm::mat4_cast(orientation->value));
-				glm::vec4 local_coords = model_view * glm::vec4(data->ray_hit_piont_world, 1.0f);
+				glm::vec4 local_coords = model_view * glm::vec4(data->ray_hit_point_world, 1.0f);
```
Comment (Meisaka) [inline client/voxel-volume.cpp:239]: spelling ;-)

File: common/components/velocity.hpp [hunk 5]
```diff
@@ -7,13 +7,13 @@
...
+		Velocity() : linear(0, 0, 0), angular(0, 0, 0) { };
 
-		Velocity(glm::vec4 linear, glm::vec4 angular)
+		Velocity(glm::vec3 linear, glm::vec3 angular)
 			: linear(linear), angular(angular) { };
 
-		glm::vec4 linear;
-		glm::vec4 angular;
+		glm::vec3 linear;
+		glm::vec3 angular;
```
Comment (Meisaka) [inline common/components/velocity.hpp:16]: I'm not sure why these were glm::vec4 when they only hold 3D vectors, if this was an early optimization, it probably wasn't a good one

File: client/game.cpp [hunk 6]
```diff
@@ -106,13 +110,16 @@ namespace tec {
...
+		// TODO: a better representation of commands so we can send them less often
+		const double COMMAND_RATE = 1.0 / 30.0;
 		delta_accumulator += delta;
-
 		game_state_queue.ProcessEventQueue();
 		game_state_queue.Interpolate(delta);
 
 		auto client_state = simulation.Simulate(delta, game_state_queue.GetInterpolatedState());
-		if (delta_accumulator >= UPDATE_RATE) {
+		game_state_queue.UpdatePredictions(client_state);
```
Comment (adam4813) [inline client/game.cpp:120]: Good idea.

File: client/game.cpp [hunk 7]
```diff
@@ -106,13 +110,16 @@ namespace tec {
 	}
 
 	void Game::Update(double delta, double mouse_x, double mouse_y, int window_width, int window_height) {
+		// TODO: a better representation of commands so we can send them less often
+		const double COMMAND_RATE = 1.0 / 30.0;
```
Comment (adam4813) [inline client/game.cpp:114]: Should this be decoupled form the update rate?

File: common/controllers/fps-controller.cpp [hunk 8]
```diff
@@ -22,10 +22,10 @@ namespace tec {
 				Handle(key_event, state);
 			}
 		}
+		for (const MouseBtnEvent& mouse_button_event : commands.mouse_button_events) {
+			Handle(mouse_button_event, state);
+		}
```
Comment (Meisaka) [inline common/controllers/fps-controller.cpp:27]: I moved this out of the mouse_focus during testing, it's probably okay until something ends up getting added that focuses the mouse, but isn't this whole class a stopgap anyways?

File: common/controllers/fps-controller.cpp [hunk 9]
```diff
@@ -195,12 +192,12 @@ namespace tec {
 		}
 
 		if (change_x != 0) {
-			glm::quat rotX = glm::angleAxis(static_cast<float>(glm::radians(change_x * -50.0f * this->current_delta)),
+			glm::quat rotX = glm::angleAxis(static_cast<float>(glm::radians(change_x * -8.0f * this->current_delta)),
```
Comment (Meisaka) [inline common/controllers/fps-controller.cpp:195]: I really had to tone these down, too flingy for me.

File: client/game.hpp [hunk 10]
```diff
@@ -1,4 +1,5 @@
-#pragma once
```
Comment (adam4813) [inline client/game.hpp:1]: I have updated the style guides to indicate using `#prgram once`. You could revert some of the changes in this PR, or it can be part of a bigger style/format PR.

File: client/gui/console.cpp [hunk 11]
```diff
@@ -221,7 +222,7 @@ namespace tec {
 	}
 
 	void ConsoleSink::log(const spdlog::details::log_msg& msg) {
-		std::string str(msg.payload.data());
+		std::string str(msg.payload.data(), msg.payload.size());
```
Comment (adam4813) [inline client/gui/console.cpp:225]: Nice optimization 👍🏻

File: common/event-queue.hpp [hunk 12]
```diff
@@ -1,60 +1,79 @@
...
+	}
+	virtual ~EventQueue() {}
 
-		void QueueEvent(Event<T>&& e) {
-			(*write_event_queue).emplace(std::move(e));
+	void ProcessEventQueue() {
+		this->read_event_queue = write_event_queue.exchange(this->read_event_queue);
+		std::scoped_lock<std::mutex> lock(this->read_event_queue->queue_write);
+		while (!this->read_event_queue->queue.empty()) {
+			Event<T> e = std::move(this->read_event_queue->queue.front());
```
Comment (Meisaka) [inline common/event-queue.hpp:60]: since so many lines in this file were being changed to add the extra wrapper class, I just went ahead and changed the namespace indent

File: client/imgui-system.cpp [hunk 13]
```diff
@@ -369,7 +377,15 @@ namespace tec {
 	}
 
 	void IMGUISystem::On(std::shared_ptr<MouseBtnEvent> data) {
-		this->mouse_pressed[data->button] = data->action == MouseBtnEvent::DOWN;
+		if (data->action == MouseBtnEvent::DOWN) {
+			this->mouse_pressed[data->button] = true;
+		}
+		// Imgui processes mouse events per frame, so if a click lasts less than a frame could be missed
```
Comment (adam4813) [inline client/imgui-system.cpp:383]: Very useful comment BTW.

File: common/game-state.hpp [hunk 14]
```diff
@@ -115,6 +121,7 @@ namespace tec {
 		}
 		state_id_t state_id = 0;
 		state_id_t command_id = 0;
+		uint64_t timestamp = 0;
```
Comment (Meisaka) [inline common/game-state.hpp:124]: this could in-theory be a uint32_t, since we don't care if it wraps or not

File: common/physics-system.cpp [hunk 15]
```diff
@@ -67,14 +67,22 @@ namespace tec {
...
 					collidable->motion_state.transform.setOrigin(btVector3(position.x, position.y, position.z));
 				}
 			}
-			if (state.orientations.find(entity_id) != state.orientations.end()) {
-				glm::quat orientation = state.orientations.at(entity_id).value;
+			else {
+				// no position! that's not good
+				// wait to add the physics body to the world for now
+				continue;
+			}
```
Comment (Meisaka) [inline common/physics-system.cpp:82]: This is to try and work around some race condition between positions being set, it doesn't fix the race though

File: client/os.cpp [hunk 16]
```diff
@@ -424,9 +464,10 @@ namespace tec {
 		if (action == GLFW_PRESS) {
 			mbtn_event->action = MouseBtnEvent::DOWN;
 		}
-		else {
+		else if(action == GLFW_RELEASE) {
 			mbtn_event->action = MouseBtnEvent::UP;
 		}
+		else return;
```
Comment (adam4813) [inline client/os.cpp:470]: Without the bracing this is a bit hard to see since the other branches use braces.

File: client/os.cpp [hunk 17]
```diff
@@ -397,6 +404,39 @@ namespace tec {
...
+					static_cast<int>(x),
+					static_cast<int>(y)
+				});
+			EventSystem<MouseMoveEvent>::Get()->Emit(mmov_event);
+			double client_center_x = static_cast<double>(this->client_width / 2);
+			double client_center_y = static_cast<double>(this->client_height / 2);
+			// constrain the mouse towards the center of the window
+			// we do this to prevent an initial frame of "overshoot"
+			// and also to prevent anything using absolute mode from wandering away, i.e. tooltip window
+			if (x < client_center_x && x > this->old_mouse_x) {
```
Comment (adam4813) [inline client/os.cpp:425]: This could a great help function `bool IsBetween(T val, T lhs, T rhs)` and I am sure it could be useful in other locations.

File: client/os.cpp [hunk 18]
```diff
@@ -260,6 +263,10 @@ namespace tec {
 	void OS::OSMessageLoop() {
 		glfwPollEvents();
 		EventQueue<KeyboardEvent>::ProcessEventQueue();
+		if (this->mouse_lock != OS::mouse_locked) {
```
Comment (adam4813) [inline client/os.cpp:266]: Is there a reason to do this check every frame instead of via the enable/disable methods?

File: client/server-connection.cpp [hunk 19]
```diff
@@ -170,9 +170,12 @@ namespace tec {
 			}
 		}
 
-		void ServerConnection::SyncHandler(const ServerMessage&) {
+		void ServerConnection::SyncHandler(const ServerMessage& message) {
 			std::chrono::milliseconds round_trip = std::chrono::duration_cast<std::chrono::milliseconds>(recv_time - sync_start);
 			std::lock_guard<std::mutex> recent_ping_lock(recent_ping_mutex);
+			if(message.GetBodyLength() >= 8) {
```
Comment (adam4813) [inline client/server-connection.cpp:176]: What does 8 represent?

Speaking of the 10 constant just below this should be swapped for a constant.

File: common/physics-system.cpp [hunk 20]
```diff
@@ -93,21 +102,49 @@ namespace tec {
...
+			else {
+				btTransform& body_transform = body->getWorldTransform();
+				// simulation motion estimation lite
+				// on the server, this doesn't really do anything
+				// on the client however, this smooths out the motion between the local estimation and server state
+				// this provides a crude but effective error correction until a better one is added ;)
+				if (body_transform.getOrigin().distance(collidable->motion_state.transform.getOrigin()) > 0.01) {
+					// FIXME there is a race condition outside of PhysicsSystem, where the position of the entity
+					// is momentarily at origin (0,0,0) during entity creation (affects server)
+					body->translate(0.5 * (collidable->motion_state.transform.getOrigin() - body_transform.getOrigin()));
```
Comment (Meisaka) [inline common/physics-system.cpp:136]: the 0.5 is the error correction curve factor, I'm indecisive about it, but 0.5 seems like a decent compromise, smaller numbers are smoother, but also make controls feel more "floaty". Simply snapping to the error target doesn't always work great either.

File: common/physics-system.cpp [hunk 21]
```diff
@@ -117,8 +154,11 @@ namespace tec {
 			}
 		}
 
-		this->dynamicsWorld->stepSimulation(static_cast<btScalar>(delta), 10);
+		// using a delta time here makes physics far less deterministic
+		// this can be changed if it becomes a problem
+		this->dynamicsWorld->stepSimulation(static_cast<btScalar>(delta), this->simulation_substeps);
```
Comment (Meisaka) [inline common/physics-system.cpp:159]: This was made configurable, seemed like a good idea

File: common/CMakeLists.txt [hunk 22]
```diff
@@ -38,7 +37,22 @@ set(trillek-common_SOURCES
...
+set_property(SOURCE ${trillek-common_PROTOCOLS} PROPERTY HEADER_FILE_ONLY true)
+
+add_library(${trillek-common_LIBRARY} STATIC
+	${COMPONENTS_SOURCES}
+	${GAME_STATE_SOURCES}
+	${COMMANDS_SOURCES}
+	${trillek-common_SOURCES}
+	${trillek-common_HEADERS}
+	${trillek-common_PROTOCOLS}
+	)
```
Comment (adam4813) [inline common/CMakeLists.txt:55]: Odd indent

File: common/physics-system.cpp [hunk 23]
```diff
@@ -331,30 +379,19 @@ namespace tec {
...
-				case proto::Component::kVelocity:
-				case proto::Component::kAudioSource:
-				case proto::Component::kPointLight:
-				case proto::Component::kDirectionalLight:
-				case proto::Component::kSpotLight:
-				case proto::Component::kVoxelVolume:
-				case proto::Component::kComputer:
-				case proto::Component::kLuaScript:
-				case proto::Component::COMPONENT_NOT_SET:
+			default:
```
Comment (Meisaka) [inline common/physics-system.cpp:394]: "default" is a thing :)

File: common/event-queue.hpp [hunk 24]
```diff
@@ -1,60 +1,79 @@
-#pragma once
+#ifndef TRILLEK_COMMON_EVENT_QUEUE_HPP
```
Comment (adam4813) [inline common/event-queue.hpp:1]: See comment above about updating style guide to prefer `#pragma once`

File: common/server-stats.hpp [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+#include <memory>
+#include <atomic>
+#include <chrono>
+
+#include <glm/glm.hpp>
+#include "tec-types.hpp"
+
+namespace tec {
+
+class ServerStats {
```
Comment (adam4813) [inline common/server-stats.hpp:14]: Could this just be a struct and remove the source file?

File: common/physics-system.cpp [hunk 26]
```diff
@@ -367,18 +404,22 @@ namespace tec {
 	}
 
 	Position PhysicsSystem::GetPosition(eid entity_id) {
-		if (this->bodies.find(entity_id) != this->bodies.end() && this->bodies.at(entity_id)) {
-			auto pos = static_cast<CollisionBody*>(this->bodies.at(entity_id)->getUserPointer())->motion_state.transform.getOrigin();
+		auto body_iter = this->bodies.find(entity_id);
+		if (body_iter != this->bodies.end() && body_iter->second) {
+			//CollisionBody* body = static_cast<CollisionBody*>(body_iter->second->getUserPointer());
+			auto pos = body_iter->second->getWorldTransform().getOrigin();
```
Comment (Meisaka) [inline common/physics-system.cpp:410]: Changed this while debugging why bullet was giving me weird values, and then not giving me values. Turns out it's related to the simulation substeps setting, I left them this way because the game state pulls the transforms from here, and we want the non-interpolated versions, where as the custom motion state object gets the interpolated transform.
tl;dr: Feeding the MotionState transform back into Physics was breaking things

File: common/proto/game_state.proto [hunk 27]
```diff
@@ -7,5 +7,6 @@ import "components.proto";
 message GameStateUpdate {
 	required uint64 state_id = 1;
 	required uint64 command_id = 2;
-	repeated Entity entity = 3;
-}
\ No newline at end of file
+	required uint64 timestamp = 3;
+	repeated Entity entity = 7;
```
Comment (Meisaka) [inline common/proto/game_state.proto:11]: Since I wanted to move this, I left some room

File: common/server-stats.cpp [hunk 28]
```diff
@@ -0,0 +1,7 @@
+
+#include "server-stats.hpp"
+
+namespace tec {
+namespace networking {
+}
+}
```
Comment (Meisaka) [inline common/server-stats.cpp:7]: yes, this file does nothing, yes it's still here anyways.

File: server/CMakeLists.txt [hunk 29]
```diff
@@ -17,11 +17,21 @@ endif ()
...
+# Used by IDEs to nicely view header files, don't depend on this to work!
+file(GLOB_RECURSE trillek-server_HEADERS LIST_DIRECTORIES false ${trillek-server_SOURCE_DIR}/*.hpp)
+source_group(Headers FILES ${trillek-server_HEADERS})
+set_property(SOURCE ${trillek-server_HEADERS} PROPERTY HEADER_FILE_ONLY true)
+
+add_executable(${trillek-server_PROGRAM}
+	${trillek-server_SOURCE_DIR}/main.cpp
+	${trillek-server_SOURCES}
+	${trillek-server_HEADERS}
+	)
```
Comment (adam4813) [inline server/CMakeLists.txt:34]: Odd indent

File: server/main.cpp [hunk 30]
```diff
@@ -57,10 +60,16 @@ int main() {
 
 	// Accumulated deltas since the last update was sent.
 	double delta_accumulator = 0.0;
+	// Accumulated deltas since the last physics simulation.
```
Comment (adam4813) [inline server/main.cpp:63]: Copy/paste comment miss?

File: common/simulation.cpp [hunk 31]
```diff
@@ -11,7 +11,8 @@
 #include "controllers/fps-controller.hpp"
 
 namespace tec {
-	double UPDATE_RATE = 10.0 / 60.0;
+
+	double UPDATE_RATE = 1.0 / 8.0; // 8 per second
```
Comment (Meisaka) [inline common/simulation.cpp:15]: Slight increase from 6 per second!

File: common/simulation.hpp [hunk 32]
```diff
@@ -1,4 +1,5 @@
-#pragma once
+#ifndef TRILLEK_COMMON_SIMULATION_HPP
+#define TRILLEK_COMMON_SIMULATION_HPP
```
Comment (Meisaka) [inline common/simulation.hpp:2]: I'm unsure if this provides me any benefit or not

File: server/client-connection.cpp [hunk 33]
```diff
@@ -132,48 +132,55 @@ namespace tec {
...
-							case MessageType::GAME_STATE_UPDATE:
-								break;
-						}
-						std::this_thread::sleep_for(std::chrono::milliseconds(1));
-						read_header();
-					}
-					else {
+					auto now_time = std::chrono::high_resolution_clock::now();
+					uint64_t current_timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(now_time.time_since_epoch()).count();
+					if (error) {
```
Comment (Meisaka) [inline server/client-connection.cpp:137]: changed to be an early-out, to reduce indentation

File: server/main.cpp [hunk 34]
```diff
@@ -73,42 +82,54 @@ int main() {
...
-							std::lock_guard lg(server.client_list_mutex);
-							for (std::shared_ptr<tec::networking::ClientConnection> client : server.GetClients()) {
-								client->UpdateGameState(full_state);
-								if (current_state_id - client->GetLastConfirmedStateID() > tec::TICKS_PER_SECOND * 2.0) {
-									server.Deliver(client, full_state_update_message);
-									std::cout << "sending full state " << current_state_id << " to: " << client->GetID() << " client state ID was: " << client->GetLastConfirmedStateID() << std::endl;
-								}
-								else {
-									server.Deliver(client, client->PrepareGameStateUpdateMessage(current_state_id));
+						tec::GameState full_state = simulation.Simulate(SERVER_SIMULATE_RATE, game_state_queue.GetBaseState());
```
Comment (Meisaka) [inline server/main.cpp:95]: I changed the server to use a fixed update rate, but I had to move things into an "if" block as result.
Depending how loaded the server is, this might make it spend more time in thread sleep.

File: common/physics-system.cpp [hunk 35]
```diff
@@ -93,21 +102,49 @@ namespace tec {
...
 			if (collidable->disable_deactivation) {
 				body->forceActivationState(DISABLE_DEACTIVATION);
 			}
 			else if (body->getActivationState() == DISABLE_DEACTIVATION) {
 				body->forceActivationState(ACTIVE_TAG);
 			}
 
+			// prevent the simulation from rotating the object
+			// this doesn't account for change after creation, once disabled there isn't a re-enable
 			if (collidable->disable_rotation) {
```
Comment (adam4813) [inline common/physics-system.cpp:115]: You must remove the rigidbody IIRC to make any of these flag updates take affect.

File: client/client-game-state-queue.cpp [hunk 36]
```diff
@@ -0,0 +1,233 @@
+#include "client-game-state-queue.hpp"
```
Comment (adam4813) [inline client/client-game-state-queue.cpp:1]: The indent on this file is 2 spaces, per the style guide it should be 4

File: client/client-game-state-queue.hpp [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+	GameState* GetGameState(int offset) {
+		return &this->server_states_array[(server_state_array_index - offset) % SERVER_STATES_ARRAY_SIZE];
+	}
+private:
+	static const unsigned int SERVER_STATES_ARRAY_SIZE{ 5 };
+
+	GameState server_states_array[SERVER_STATES_ARRAY_SIZE];
+	int server_state_array_index{ SERVER_STATES_ARRAY_SIZE - 1 };
+	GameState predicted_states[SERVER_STATES_ARRAY_SIZE];
+	int predicted_states_array_index{ 0 };
```
Comment (Meisaka) [inline client/client-game-state-queue.hpp:70]: I don't think these are actually used anywhere

File: client/client-game-state-queue.cpp [hunk 38]
```diff
@@ -0,0 +1,233 @@
...
+				glm::vec3 predict_vel = itr->second.velocities[this->client_id].linear;
+				velocity_diff = state_vel - predict_vel;
+			}
+			itr = this->predictions.erase(itr);
+			continue;
+		}
+		++itr;
+	}
+	// debugging stats, also used by reconciliation
+	stats.client_position = position_diff;
```
Comment (adam4813) [inline client/client-game-state-queue.cpp:168]: The client_position variable name is a bit misleading

File: client/client-game-state-queue.hpp [hunk 39]
```diff
@@ -0,0 +1,86 @@
...
+	ServerStats& stats;
+	GameState base_state;
+	GameState interpolated_state;
+	std::queue<GameState> server_states;
+	std::mutex server_state_mutex;
+	state_id_t last_server_state_id{ 0 };
+	state_id_t command_id{ 0 };
+	double interpolation_accumulator{ 0.0 };
+	eid client_id{ 0 };
+	std::map<state_id_t, GameState> predictions;
```
Comment (Meisaka) [inline client/client-game-state-queue.hpp:81]: predictions were made into GameState instead of just a position. However, it only ever holds the state of one entity (the client entity)

**Review by adam4813 — APPROVED** [review]
Just some minor questions.

Feel free to address style/format issues if you want.

File: client/gui/console.cpp (hunk 11 above)
Comment (Meisaka) [reply client/gui/console.cpp:225]: this is actually a bug fix to prevent it from spewing garbage into the console

Comment (adam4813) [reply client/gui/console.cpp:225]: Oh I tried to fix that one day, glad you figured it out.

File: client/server-connection.cpp (hunk 19 above)
Comment (Meisaka) [reply client/server-connection.cpp:176]: probably the low levelness getting the better of me, `sizeof(uint64_t)` would make more sense.

File: client/client-game-state-queue.cpp (hunk 38 above)
Comment (Meisaka) [reply client/client-game-state-queue.cpp:168]: After having refactored this a few times, I would agree.

File: client/client-game-state-queue.cpp (hunk 36 above)
Comment (Meisaka) [reply client/client-game-state-queue.cpp:1]: not sure I follow...

File: client/game.cpp (hunk 7 above)
Comment (Meisaka) [reply client/game.cpp:114]: Ideally, there would be one command per frame, but instead of sending them that frequently, they get batched up with a timestamp or something and sent in bursts. This should be decoupled from the update rate, and set fairly high to maintain smooth control, but since we don't have batching, this is the compromise setting

File: client/client-game-state-queue.cpp (hunk 36 above)
Comment (Meisaka) [reply client/client-game-state-queue.cpp:1]: if you are referring to how it's rendered on Github, that might just be a browser display issue, since all these files use TAB

File: client/os.cpp (hunk 18 above)
Comment (Meisaka) [reply client/os.cpp:266]: The enable/disable methods could be called from any thread, but the `glfwSetInputMode` call has to be on this thread or it doesn't work.

**Review by adam4813 — APPROVED** [review]
